# Analysing the difference of timestamps of buoys

in 1_buoy_data_analysis.ipynb, I have observed that the timestamps of different buoys do not match. That is a problem since this does not allow me to create one instance with values for all buoys. In this notebook, I am going to observe the problem in more detail and might come to a solution.


In [15]:
import pandas as pd
from collections import Counter
from IPython.core.display_functions import display

import myLibrary as mL

### Stations for gulf of mexico
copied from: https://www.ndbc.noaa.gov/mobile/region.php?reg=gulf_of_mexico

# Test which buoys actually exist!

In [16]:
# Clean data: excluded to save computation time, see hardcopied values below!
cleaned_stations_GOM, failed_stations_GOM = mL.clean_station_list(mL.stations_GOM, "2020")
display(cleaned_stations_GOM)
display(failed_stations_GOM)

Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404: Not Found
Failed to get file: HTTP Error 404

['41117',
 '41112',
 '42001',
 '42002',
 '42012',
 '42013',
 '42019',
 '42020',
 '42022',
 '42023',
 '42026',
 '42036',
 '42039',
 '42040',
 '42055',
 '42095',
 '42097',
 '42098',
 '42099',
 'amrl1',
 'anpt2',
 'apcf1',
 'arpf1',
 'awrt2',
 'babt2',
 'bktl1',
 'bsca1',
 'burl1',
 'bygl1',
 'bzst2',
 'capl1',
 'carl1',
 'cdrf1',
 'cnbf1',
 'crta1',
 'cwaf1',
 'cwbf1',
 'dmsf1',
 'ebef1',
 'einl1',
 'emat2',
 'eptt2',
 'fhpf1',
 'fmoa1',
 'fmrf1',
 'fpst2',
 'frdf1',
 'gbif1',
 'gctf1',
 'gisl1',
 'gkyf1',
 'gnjt2',
 'grrt2',
 'gtot2',
 'hist2',
 'href1',
 'irdt2',
 'jxuf1',
 'ktnf1',
 'kywf1',
 'lcll1',
 'lmrf1',
 'ltjf1',
 'luit2',
 'mbet2',
 'mbpa1',
 'mcga1',
 'mgpt2',
 'mhbt2',
 'mtbf1',
 'mypf1',
 'ncht2',
 'nfdf1',
 'nuet2',
 'nwcl1',
 'obla1',
 'optf1',
 'pacf1',
 'pact2',
 'pcbf1',
 'pcgt2',
 'pclf1',
 'pcnt2',
 'pill1',
 'pmaf1',
 'pmnt2',
 'pnlm6',
 'port2',
 'pstl1',
 'ptat2',
 'ptbm6',
 'ptit2',
 'ptoa1',
 'rcpt2',
 'rlit2',
 'rlot2',
 'rsjt2',
 'rtat2',
 'sapf1',
 'sauf1',


['42084',
 '42091',
 'apxf1',
 'bkbf1',
 'dila1',
 'efla1',
 'gtxf1',
 'hivt2',
 'katp',
 'kbmg1',
 'kbqx',
 'kdlp',
 'kgry',
 'kgul',
 'khhv',
 'kikt',
 'kvaf',
 'kvoa',
 'lqat2',
 'maxt2',
 'rkxf1',
 'vtbt2',
 'wkxa1']

In [17]:
print(len(mL.stations_GOM), "Stations in total in the Gulf of Mexico.")
print(len(cleaned_stations_GOM)," stations provide data for 2020.")
print(len(failed_stations_GOM)," stations do not provide data for year 2020.")

146 Stations in total in the Gulf of Mexico.
123  stations provide data for 2020.
23  stations do not provide data for year 2020.


# Compare the timestamps of the data

In [18]:
def create_timestamp_df(station_numbers, year):
    #time_df = pd.DataFrame()
    time_dict = dict()
    for station_number in station_numbers:
        data = mL.get_data_file(station_number, year)  #load file

        if data is not None:
            filename = data[0]
            station_df = data[1].drop(labels=0, axis=0) #drop row with units
            #print(station_df.shape)

            time_dict[filename] =  station_df['#YY'] +"-" \
                            + station_df['MM'] +"-"  \
                            + station_df['DD'] +" " \
                            + station_df['hh'] +":"\
                            + station_df['mm']

            #TODO: following line causes performance warning!
            #Is there a way to optimize this, doing the whole joining in one step?
            time_dict[filename] = pd.to_datetime(time_dict[filename], format='%Y-%m-%d %H:%M', utc=True)
            #time_df[filename] = time_df['Formatted Date'].apply(pd.to_datetime, utc=True)

    return pd.DataFrame(time_dict)

In [19]:
timestamp_df = create_timestamp_df(cleaned_stations_GOM, "2020")
timestamp_df

41117                     41112  \
1     2020-01-01 00:00:00+00:00 2020-01-01 00:00:00+00:00   
2     2020-01-01 00:30:00+00:00 2020-01-01 00:30:00+00:00   
3     2020-01-01 01:00:00+00:00 2020-01-01 01:00:00+00:00   
4     2020-01-01 01:30:00+00:00 2020-01-01 01:30:00+00:00   
5     2020-01-01 02:00:00+00:00 2020-01-01 02:00:00+00:00   
...                         ...                       ...   
87263                       NaT                       NaT   
87264                       NaT                       NaT   
87265                       NaT                       NaT   
87266                       NaT                       NaT   
87267                       NaT                       NaT   

                          42001                     42002  \
1     2020-01-01 00:00:00+00:00 2020-01-01 00:00:00+00:00   
2     2020-01-01 00:10:00+00:00 2020-01-01 00:10:00+00:00   
3     2020-01-01 00:20:00+00:00 2020-01-01 00:20:00+00:00   
4     2020-01-01 00:30:00+00:00 2020-01-01 00:30:00+00:00   
5     2020-01-01 00:40:00+00:00 2020-01-01 00:40:00+00:00   
...                         ...                       ...   
87263                       NaT                       NaT   
87264                       NaT                       NaT   
87265                       NaT                       NaT   
87266                       NaT                       NaT   
87267                       NaT                       NaT   

                          42012                     42013  \
1     2020-01-01 00:00:00+00:00 2020-01-30 00:00:00+00:00   
2     2020-01-01 00:10:00+00:00 2020-01-30 00:30:00+00:00   
3     2020-01-01 00:20:00+00:00 2020-01-30 01:00:00+00:00   
4     2020-01-01 00:30:00+00:00 2020-01-30 01:30:00+00:00   
5     2020-01-01 00:40:00+00:00 2020-01-30 02:00:00+00:00   
...                         ...                       ...   
87263                       NaT                       NaT   
87264                       NaT                       NaT   
87265                       NaT                       NaT   
87266                       NaT                       NaT   
87267                       NaT                       NaT   

                          42019                     42020  \
1     2020-01-01 00:00:00+00:00 2020-01-01 00:00:00+00:00   
2     2020-01-01 00:10:00+00:00 2020-01-01 00:10:00+00:00   
3     2020-01-01 00:20:00+00:00 2020-01-01 00:20:00+00:00   
4     2020-01-01 00:30:00+00:00 2020-01-01 00:30:00+00:00   
5     2020-01-01 00:40:00+00:00 2020-01-01 00:40:00+00:00   
...                         ...                       ...   
87263                       NaT                       NaT   
87264                       NaT                       NaT   
87265                       NaT                       NaT   
87266                       NaT                       NaT   
87267                       NaT                       NaT   

                          42022                     42023  ...  \
1     2020-01-01 01:00:00+00:00 2020-01-01 00:00:00+00:00  ...   
2     2020-01-01 01:30:00+00:00 2020-01-01 00:30:00+00:00  ...   
3     2020-01-01 02:00:00+00:00 2020-01-01 01:00:00+00:00  ...   
4     2020-01-01 02:30:00+00:00 2020-01-01 01:30:00+00:00  ...   
5     2020-01-01 03:00:00+00:00 2020-01-01 02:00:00+00:00  ...   
...                         ...                       ...  ...   
87263                       NaT                       NaT  ...   
87264                       NaT                       NaT  ...   
87265                       NaT                       NaT  ...   
87266                       NaT                       NaT  ...   
87267                       NaT                       NaT  ...   

                          tshf1                     txpt2  \
1     2020-01-01 00:00:00+00:00 2020-01-01 00:00:00+00:00   
2     2020-01-01 00:06:00+00:00 2020-01-01 00:06:00+00:00   
3     2020-01-01 00:12:00+00:00 2020-01-01 00:12:00+00:00   
4     2020-01-01 00:18:00+00:00 2020-01-01 00:18:00+00:00

### Analysis Step 1:
Do all buoys start at 2020-01-01 00:00?

Note: +00:00 is a UTC offset - thus can be interpreted as UTC (=Universal Time Coordinated)

In [20]:
def isSameFirstValue(df):
    first_row = df.head(1)
    first_timestamp = first_row.iloc[0, 0]
    print("First timestamp: ", first_timestamp)

    is_same = True
    counter = 0
    for column in first_row:
        if first_timestamp != first_row.at[1, column]:
            print("Different Value in ", column, ": ", first_row.at[1, column])
            counter = counter + 1
            is_same = False

    print("Found ", counter, "different values!")
    return is_same

print(isSameFirstValue(timestamp_df))

First timestamp:  2020-01-01 00:00:00+00:00
Different Value in  42013 :  2020-01-30 00:00:00+00:00
Different Value in  42022 :  2020-01-01 01:00:00+00:00
Different Value in  42055 :  2020-01-01 00:40:00+00:00
Different Value in  bsca1 :  2020-02-01 00:00:00+00:00
Different Value in  ebef1 :  2020-05-01 00:00:00+00:00
Different Value in  irdt2 :  2020-01-02 14:00:00+00:00
Different Value in  jxuf1 :  2020-05-06 10:42:00+00:00
Different Value in  mhbt2 :  2020-05-26 13:54:00+00:00
Different Value in  rlit2 :  2020-03-03 08:18:00+00:00
Different Value in  skcf1 :  2020-03-10 21:00:00+00:00
Different Value in  tlvt2 :  2020-12-23 14:30:00+00:00
Different Value in  txvt2 :  2020-11-19 18:42:00+00:00
Different Value in  utvt2 :  2020-11-23 14:54:00+00:00
Found  13 different values!
False


## ISSUE: not all Buoys start on the same time


### Analysis Step 2:
What are the time-steps?

Calculate the (average?!) time difference between the timestamps

In [21]:
#Param: timestamp: pandas.core.series.Series (of a single buoy)
#Return: dataframe with time-delta as index and count value in first column. the columns name is the buoy-id
def calc_time_delta_counts(timestamps, heading):
    # build a list of all time steps for a specific buoy and their counts
    # calc average time stamps

    #display(type(timestamps))
    timedeltas = [timestamps[i]-timestamps[i-1] for i in range(2, len(timestamps))]
    counts = dict(Counter(timedeltas))
    counts_df = pd.DataFrame.from_dict(counts, orient='index').reset_index()
    counts_df.columns = ["timedelta",heading]
    counts_df.set_index("timedelta", inplace = True)
    #display(counts_df)

    # average = pd.Series(timedeltas).sum() / len(timedeltas)
    # print("Average timedelta: ", average)
    counts_df.drop("NaT", inplace=True, errors='ignore')    #errors='ignore' only drops row if it exists

    return counts_df

#TEST:
col_name = timestamp_df.columns[0]
calc_time_delta_counts(timestamp_df[col_name], col_name)

41117
timedelta             
0 days 00:30:00  17345
0 days 01:00:00     54
0 days 01:30:00      4
0 days 07:30:00      1
0 days 00:26:00      1
0 days 02:00:00      2
0 days 02:30:00      1
0 days 22:00:00      1
0 days 11:30:00      1
0 days 03:30:00      1

In [22]:
#Param: timestamps: dataframe with timestamps as it is returned by create_timestamp_df()
def calc_all_time_delta_counts(timestamps):
    counts_df = pd.DataFrame()

    for buoy_id in timestamps:
        current_counts = calc_time_delta_counts(timestamps[buoy_id], heading=buoy_id)
        counts_df = pd.concat([counts_df, current_counts], axis=1)
        #counts_df[buoy_id] = current_counts
        counts_df.fillna(0, inplace = True)
    return counts_df.sort_index()

In [23]:
timestep_counts = calc_all_time_delta_counts(timestamp_df)
timestep_counts

41117  41112    42001    42002    42012  42013    42019  \
timedelta                                                                     
-13 days +09:30:00    0.0    0.0      0.0      0.0      0.0    0.0      0.0   
0 days 00:01:00       0.0    0.0      0.0      0.0      0.0    0.0      0.0   
0 days 00:05:00       0.0    0.0      0.0      0.0      0.0    0.0      0.0   
0 days 00:06:00       0.0    0.0      0.0      0.0      0.0    0.0      0.0   
0 days 00:10:00       0.0    0.0  37464.0  52173.0  52146.0    0.0  46427.0   
...                   ...    ...      ...      ...      ...    ...      ...   
55 days 06:06:00      0.0    0.0      0.0      0.0      0.0    0.0      0.0   
56 days 00:18:00      0.0    0.0      0.0      0.0      0.0    0.0      0.0   
89 days 05:30:00      0.0    0.0      0.0      0.0      0.0    0.0      0.0   
156 days 09:00:00     0.0    0.0      0.0      0.0      0.0    0.0      0.0   
227 days 13:06:00     0.0    0.0      0.0      0.0      0.0    0.0      0.0   

                      42020  42022  42023  ...    tshf1    txpt2    txvt2  \
timedelta                                  ...                              
-13 days +09:30:00      0.0    0.0    0.0  ...      0.0      0.0      0.0   
0 days 00:01:00         0.0    0.0    0.0  ...      0.0      0.0      0.0   
0 days 00:05:00         0.0    0.0    0.0  ...      0.0      0.0      0.0   
0 days 00:06:00         0.0    0.0    0.0  ...  80600.0  83265.0  10087.0   
0 days 00:10:00     30464.0    1.0    1.0  ...      0.0      0.0      0.0   
...                     ...    ...    ...  ...      ...      ...      ...   
55 days 06:06:00        0.0    0.0    0.0  ...      0.0      0.0      0.0   
56 days 00:18:00        0.0    0.0    0.0  ...      0.0      0.0      0.0   
89 days 05:30:00        0.0    0.0    0.0  ...      0.0      0.0      0.0   
156 days 09:00:00       0.0    0.0    0.0  ...      0.0      0.0      0.0   
227 days 13:06:00       0.0    0.0    0.0  ...      0.0      0.0      0.0   

                     utvt2    vcaf1    vcat2  venf1    wiwf1    wplf1    wycm6  
timedelta                                                                       
-13 days +09:30:00     0.0      0.0      0.0    0.0      0.0      0.0      0.0  
0 days 00:01:00        0.0      0.0      0.0    0.0      0.0      0.0      0.0  
0 days 00:05:00        0.0      0.0      0.0    0.0      0.0      0.0      0.0  
0 days 00:06:00     8949.0  86639.0  85969.0    0.0  79270.0  81253.0  80115.0  
0 days 00:10:00        0.0      0.0      0.0    0.0      0.0      0.0      0.0  
...                    ...      ...      ...    ...      ...      ...      ...  
55 days 06:06:00       0.0      0.0      0.0    0.0      0.0      0.0      0.0  
56 days 00:18:00       0.0      0.0      0.0    0.0      0.0      0.0      0.0  
89 days 05:30:00       0.0      0.0      0.0    0.0      0.0      0.0      0.0  
156 days 09:00:00      0.0      0.0      0.0    0.0      0.0      0.0      0.0  
227 days 13:06:00      0.0      0.0      0.0    0.0      0.0      0.0      0.0  

[298 rows x 123 columns]

as we can see, we do have 298 different time_steps ...
now I need to find a way to get as much data as possible but with consistent timesteps!

In [24]:
def create_max_count_statistic(timestep_counts):

    max_counts = pd.DataFrame() #statistic by buoy ID
    max_counts["max_count"] = timestep_counts.max()
    max_counts["timedelta"] = timestep_counts.idxmax()
    max_counts["num_of_records"] = timestep_counts.sum()

    #total_sum = timestep_counts.columns[1].sum()
    #print("total sum: ", total_sum)

    max_counts["ratio in %"] = max_counts["max_count"] / max_counts["num_of_records"]
    #print(type(max_counts))
    #display(max_counts)

    max_counts_by_delta = pd.DataFrame()    #statistic by timedelta
    max_counts_by_delta["timedelta"] = max_counts["timedelta"].unique()
    #display(max_counts_by_delta)
    #print("Type of max_counts_by_delta: ", type(max_counts_by_delta))

    buoy_ids = list()   #list of all timestamps
    summs = list()        #total sum of timedelta values
    for index_1, row_1 in max_counts_by_delta.iterrows():
        current_ids = list() #this list becomes an element of buoy_ids
        current_sum = 0
        for index_2, row_2 in max_counts.iterrows():
            if row_1["timedelta"] == row_2["timedelta"]:
                current_ids.append(index_2)
                current_sum += row_2["max_count"]
        buoy_ids.append(current_ids)
        summs.append(current_sum)

    max_counts_by_delta["buoy_ids"] = buoy_ids
    max_counts_by_delta["#total_obersvations"] = summs

    max_counts_by_delta.set_index('timedelta', inplace = True)
    return max_counts, max_counts_by_delta

#TEST:
print(type(timestep_counts))
statistic_by_buoy_id, statistic_by_timedelta = create_max_count_statistic(timestep_counts)
statistic_by_buoy_id

<class 'pandas.core.frame.DataFrame'>


max_count       timedelta  num_of_records  ratio in %
41117    17345.0 0 days 00:30:00         17411.0    0.996209
41112    15820.0 0 days 00:30:00         15909.0    0.994406
42001    37464.0 0 days 00:10:00         37578.0    0.996966
42002    52173.0 0 days 00:10:00         52272.0    0.998106
42012    52146.0 0 days 00:10:00         52258.0    0.997857
...          ...             ...             ...         ...
vcat2    85969.0 0 days 00:06:00         86191.0    0.997424
venf1     8745.0 0 days 01:00:00          8757.0    0.998630
wiwf1    79270.0 0 days 00:06:00         79924.0    0.991817
wplf1    81253.0 0 days 00:06:00         81779.0    0.993568
wycm6    80115.0 0 days 00:06:00         83079.0    0.964323

[123 rows x 4 columns]

In [25]:
statistic_by_timedelta

buoy_ids  \
timedelta                                                            
0 days 00:30:00  [41117, 41112, 42013, 42022, 42023, 42026, 420...   
0 days 00:10:00  [42001, 42002, 42012, 42019, 42020, 42036, 420...   
0 days 00:06:00  [amrl1, anpt2, apcf1, arpf1, awrt2, babt2, bkt...   
0 days 01:00:00  [burl1, cdrf1, ktnf1, ptat2, sauf1, sgof1, srs...   

                 #total_obersvations  
timedelta                             
0 days 00:30:00             177487.0  
0 days 00:10:00             432007.0  
0 days 00:06:00            7395972.0  
0 days 01:00:00              63931.0

The ratio in the first statistic shows how constant the timestep is. In theory, if the sensor would save data after all x minutes, this value would be 100%.
The lowest ratio is 87%. Which means only considering the most often occuring timedelta would not reduce the number of instances significantly.

The second statistic shows how often and for which buoys the most often observed timedeltas occur. This should help to decide which timestep to use for the final dataset.

Using 1h timesteps would include:
     50%   of 30 min timedeltas
     16.6% of 10 min timedeltas
     10%   of  6 min timedeltas
    100%   of  1 h   timedeltas
    No additional NaN values would be created

Using 30 min timesteps would include:
    100%   of 30 min timedeltas
     33.3% of 10 min timedeltas
     20%   of  6 min timedeltas
     50%   of  1 h   timedeltas
   #### BUT: 50% of the data from the 1h-timedeltas would be NaN

The smaller the selected timedelta gets, the more NaN values are created. Since we would create no NaN values at all,

## Lets figure out how much data we would lose, if we only consider 1h timesteps starting on 01-01-2020 00:00.
-->3_NaN_analysis.ipynb



In [26]:
for index, row in statistic_by_timedelta.iterrows():
    print("Num of buoy_station for timestep:", index ," = ", len(row['buoy_ids']))

Num of buoy_station for timestep: 0 days 00:30:00  =  12
Num of buoy_station for timestep: 0 days 00:10:00  =  10
Num of buoy_station for timestep: 0 days 00:06:00  =  93
Num of buoy_station for timestep: 0 days 01:00:00  =  8


# ISSUE:
As we can see, most stations use the 6 min timesteps. Therefore, it we would use significantly more data by only considering the 6 min timestaps. Before I rebuild the dataset, I have to check, which timesteps are used by the ocean models ERA5 and Wave Watch III.

Acording to Pujan, 6 minutes is to little for forecasting! It should be 1 - 3 hours. (Research and ask Austin about thins!)
